teps/0003-tekton-oci-bundles.md

```diff
@@ -0,0 +1,258 @@
+marked as a `proposed` is a working document and subject to change.  You can
+denote sections that are under active debate as follows:
+
+```
+<<[UNRESOLVED optional short context or usernames ]>>
+Stuff that is being argued.
+<<[/UNRESOLVED]>>
+```
+
+When editing TEPS, aim for tightly-scoped, single-topic PRs to keep discussions
+focused.  If you disagree with what is already in a document, open a new PR
+with suggested changes.
+
+If there are new details that belong in the TEP, edit the TEP.  Once a
+feature has become "implemented", major changes should get new TEPs.
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/teps/NNNN-TEP-template/README.md).
+
+-->
+# TEP-NNNN: Tekton OCI bundles
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Contract](#contract)
+  - [User Stories (optional)](#user-stories-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposal is to be able to bundle Tasks (and Pipelines, and
+Resources, and potentially other future config objects) into an [OCI
+Artifact](https://github.com/opencontainers/artifacts), pushed to an
+image registry, and referenced from that registry.
+
+This is a TEP for the **Spec** of the Tekton OCI bundle that were
+discussed in the following docs:
+
+- [Tekton OCI Image Catalog](https://docs.google.com/document/d/1zUVrIbGZh2R9dawKQ9Hm1Cx3GevKIfOcRO3fFLdmBDc/edit#heading=h.tp9mko2koenr)
+- [Tekton OCI Image Design](https://docs.google.com/document/d/1lXF_SvLwl6OqqGy8JbpSXRj4hWJ6CSImlxlIl4V9rnM/edit?pl=1#)
+
+This will be based from the knowledge acquired on the [experimental oci
+project](https://github.com/tektoncd/experimental/tree/master/oci) we ran.
+
+## Motivation
+
+Today, TaskRuns can be defined completely (user specifies
+`.spec.taskSpec.steps` etc.), or by referencing a Task that must have
+previously been defined in the cluster's API server (user specifies
+`.spec.taskRef.name` and optionally `.namespace` -- if a Task doesn't
+exist with the specified name in the TaskRun's namespace, we check if
+it's a ClusterTask).
+
+This is overly limiting. It makes versioning hard, and it makes
+rolling out changes to Task definitions hard.
+
+We can leverage existing tooling and infrastructure used for sharing
+OCI images. An OCI image is really just a format for a set
+of binary files that can be uploaded to a registry (think DockerHub)
+with a tagged version and digest. It can easily be used to version and
+share Tasks and other Tekton resources.
+
+### Goals
+
+The goal of this TEP is to define the specification for the Tekton OCI
+bundles. The goal of the Tekton OCI bundles is to define a Spec on top
```

Comment:
Just a `Spec`? and not a full _resource_?

Reply:
The distinction is actually really important. They key use case that OCI Bundles unlocks is the ability to refer to Tasks/Pipelines outside of the cluster. This allows multiple definitions of the same-named Task/Pipeline to exist and for users to version them. Therefore, the metadata of the resource is unimportant. It is really the spec we are after as we wish to dynamically insert that at runtime.

Reply:
It means that we again will get the same problem as this PR was intended to fix https://github.com/tektoncd/pipeline/pull/2826

By providing the _full resource_ rather than a _part of a resource_ the end-user has the possibility to add *labels* and *annotations*. But this was mostly a question - it may not be relevant - but it is hard to reason since the full context (e.g. use) is not specified here.

Reply:
The `Spec` here should be rewritten as specification, it doesn't have anything to do with the `spec` field of a Tekton Resource :sweat_smile: :wink:

Reply:
I don't think the purpose of a TEP is to get too involved into the implementation but just to entertain this...there is nothing stopping the implementation of this from propagating the metadata of the task. It is possible either way (though to be clear, my preference is to not propagate labels and annotations from tasks to the pod).